### PR TITLE
feat: support reading from delta indices

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/lance/src/dataset/scanner.rs [hunk 1]
```diff
@@ -852,24 +852,24 @@ impl Scanner {
         } else {
             Arc::new(vec![])
         };
-        let knn_idx = indices.iter().find(|i| i.fields.contains(&column_id));
-        if let Some(index) = knn_idx {
+        if let Some(index) = indices.iter().find(|i| i.fields.contains(&column_id)) {
```
Comment (chebbyChefNEQ) [inline rust/lance/src/dataset/scanner.rs:855]: nit: maybe we should save this in manifest

File: rust/lance/src/dataset/scanner.rs [hunk 2]
```diff
@@ -852,24 +852,24 @@ impl Scanner {
...
             // There is an index built for the column.
             // We will use the index.
-            if let Some(rf) = q.refine_factor {
-                if rf == 0 {
-                    return Err(Error::IO {
-                        message: "Refine factor can not be zero".to_string(),
-                        location: location!(),
-                    });
-                }
+            if q.refine_factor == Some(0) {
```
Comment (chebbyChefNEQ) [inline rust/lance/src/dataset/scanner.rs:858]: nit: use `if matches!(...)`

File: rust/lance/src/dataset/scanner.rs [hunk 3]
```diff
@@ -852,24 +852,24 @@ impl Scanner {
...
+                return Err(Error::IO {
+                    message: "Refine factor can not be zero".to_string(),
+                    location: location!(),
+                });
             }
 
-            let ann_node = self.ann(q, index, filter_plan).await?; // _distance, _rowid
+            // Find all deltas with the same index name.
+            let deltas = self.dataset.load_indices_by_name(&index.name).await?;
+            let ann_node = self.ann(q, &deltas, filter_plan).await?; // _distance, _rowid
```
Comment (chebbyChefNEQ) [inline rust/lance/src/dataset/scanner.rs:867]: iiuc, this doubles ivf routing computation. For now probably okay, but we should have a ticket to dedup IVF calculation.

File: rust/lance/src/io/exec/knn.rs [hunk 4]
```diff
@@ -299,103 +298,41 @@ impl FilterLoader for SelectionVectorToPrefilter {
...
+        .zip(stream::repeat((
+            dataset.clone(),
+            pre_filter.clone(),
+            query.clone(),
+        )))
+        .map(|(idx, (ds, pre_filter, query))| async move {
+            let index = ds
+                .open_vector_index(&query.column, &idx.uuid.to_string())
+                .await?;
+            index.search(&query, pre_filter.clone()).await
```
Comment (chebbyChefNEQ) [inline rust/lance/src/io/exec/knn.rs:328]: should we put `open_vector_index` and `search` on difference threads?

File: rust/lance/src/dataset/scanner.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance/src/dataset/scanner.rs:855]: what is `this` you mentioned to?

File: rust/lance/src/io/exec/knn.rs (hunk 4 above)
Comment (eddyxu) [reply rust/lance/src/io/exec/knn.rs:328]: what is the difference? index has data dependency here. You mean like this?

```
indices = io_thread_pool.submits(ds.open_vector_index(...))
results = cpu_pool.submits(index.search(...))
```

File: rust/lance/src/dataset/scanner.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance/src/dataset/scanner.rs:858]: fixed.

File: rust/lance/src/dataset/scanner.rs (hunk 3 above)
Comment (eddyxu) [reply rust/lance/src/dataset/scanner.rs:867]: Yea we can do the IVF first and fan out the PQ search later. Lemme take a look how much work that is.

File: rust/lance/src/dataset/scanner.rs (hunk 1 above)
Comment (chebbyChefNEQ) [reply rust/lance/src/dataset/scanner.rs:855]: oh as in `column_id` -> index mapping

**Review by chebbyChefNEQ — APPROVED** [review]

**Pull request merged**